### ISSUE TITLE
screenshot without "screenshot" message feature broken

### ISSUE THREAD

**Issue by fabiangreffrath** [fabiangreffrath/crispy-doom#270]
This feature has been broken by VSYNC rendering and by the fact that we are saving the renderer's content before it has been updated with the "cleaned up" framebuffer. Even if we force rendering *now* by calling `I_FinishUpdate()' in `G_Ticker()`, the actual refreshing of the renderer happens just after the next VSYNC.

I should rather create an intermediate SDL_Surface and blit the `rgbabuffer` content into this, using the exact same scaling factors that are used by the current renderer. The results should look almost the same - though not strictly identical.

**Comment by Zodomaniac** [comment]
What is broken in this feature? I'm testing and cannot find out

**Comment by fabiangreffrath** [comment]
Oops, sorry, you are right, the description mentiones the wrong feature. I had a chicken in the oven when I reported this... ;)

**Comment by Zodomaniac** [comment]
:)
But...this isn't broken, either)))

**Comment by fabiangreffrath** [comment]
Do you have uncapped framerate enabled?

**Comment by Zodomaniac** [comment]
No, I have VSYNC on :)

**Comment by fabiangreffrath** [comment]
o_O? And there is no "screen shot" message on the second screenshot if you take two in a row?

**Comment by Zodomaniac** [comment]
No! No message
![doom0000](https://user-images.githubusercontent.com/15119473/36307433-2e6c6cec-132d-11e8-9a17-ee0ce9232db5.png)
![doom0001](https://user-images.githubusercontent.com/15119473/36307434-2e8ade02-132d-11e8-83d8-bd330ba93291.png)

**Comment by fabiangreffrath** [comment]
I don't understand it. If I press the screenshot button twice in a row, the second screenshot will have the "screen shot" message. :/

**Comment by Zodomaniac** [comment]
@JNechaevsky Come in, marine

**Comment by JNechaevsky** [comment]
Sure, will check it today. I think I have some ideas already, but to properly understand this situation first.

**Comment by JNechaevsky** [comment]
Suggestion №0: what is key a combination for clean screenshot? Is it `SHIFT` + "screen shot" key?
Suggestion №1: perhaps it should be documented in the [Wiki](https://www.chocolate-doom.org/wiki/index.php/Crispy_Doom), it's oblivious and probably not needed to keep it as undocumented feature.

**Comment by JNechaevsky** [comment]
Wait a sec. For me this feature is totally broken! Matters no if framerate is set of 35 FPS / Uncapped / VSync. Here's what is happening in 20180222-win32 build, I have binded "clean screenshot" to `/`:

- https://www.youtube.com/watch?v=TGbX_yKeXW4

Instead of making clean screenshot it reduces screen size. More of than, any menu elements will not appear after pressing `/`. Summoning @fabiangreffrath!

**Comment by Zodomaniac** [comment]
@JNechaevsky 
>Suggestion №1: perhaps it should be documented in the Wiki, it's not oblivious and probably not needed to keep it as undocumented feature.

Documented it :)
BTW, I bind screenshot to ALT and clean screenshot to CTRL.

**Comment by JNechaevsky** [comment]
> Documented it :) 'Not oblivious' - I guess you meant 'not obvious'?

Oh yeah, exactly.

Something is messed in the controls with this feature. After deleting `crispy-doom.cfg` and binding clean screenshot to `ALT`, pressing `ALT` restarts current level and menus are stopping to appear.

**Comment by Zodomaniac** [comment]
@JNechaevsky 
>After deleting crispy-doom.cfg and binding clean screenshot to ALT, pressing ALT restarts current level and menus are stopping to appear.

Check out if your OS changes the keyboard mapping to English by itself. My OS does, and yours maybe doesn't, and you have native keyboard mapping enabled and combined with Russian mapping in OS it gives some undesired effects.

If that doesn't help, maybe I'll drop in at your place this weekend to see what's happening?

**Comment by JNechaevsky** [comment]
@Zodomaniac, nope, not working. My default layout is English, and I got same issue in virtual Windows 7 and Debian, I see this in console output:

`P_SetupLevel: MAP01 (doom2.wad), HMP, Time 0:00:00, Total 0:00:00, Doom map format, normal BSP nodes.`

> If that doesn't help, maybe I'll drop in at your place this weekend to see what's happening?

Sorry, no, I'm not inviting anyone in my personal keep.

**Comment by fabiangreffrath** [comment]
@JNechaevsky Wow, this is definitely not the intended behaviour. o_O

Anyway, guys, this issue is not about the "clean screenshot" feature (as I erroneously stated at first) but about the fact that the "screen shot" message should never appear if two screenshots are taken in a row.

**Comment by JNechaevsky** [comment]
Fabian, okay, let's investigate it a little bit later then.

As about "screen shot." message - hmm, I have no any text in any screenshots despite of rendering framerate. Even by pressing binded `PRNTSCRN` key as fast as possible.

**Comment by fabiangreffrath** [comment]
> Documented it 

Since we already have a section about new cheats and new command line parameters on the WIki page, maybe it makes sense to add another section about added keys and their default bindings?

**Comment by Zodomaniac** [comment]
OK, I will))

**Comment by Zodomaniac** [comment]
Updated the Wiki with the New controls (with default bindings) section.

**Comment by fabiangreffrath** [comment]
> Updated the Wiki with the New controls (with default bindings) section.

Great, thanks!

**Comment by fabiangreffrath** [comment]
> I see this in console output:

@JNechaevsky Does it help if you disable the "reload level" key?
```
--- a/src/doom/m_menu.c
+++ b/src/doom/m_menu.c
@@ -2918,7 +2918,7 @@ boolean M_Responder (event_t* ev)
        }
         // [crispy] those two can be considered as shortcuts for the IDCLEV cheat
         // and should be treated as such, i.e. add "if (!netgame)"
-        else if (!netgame && key == key_menu_reloadlevel)
+        else if (false && !netgame && key == key_menu_reloadlevel)
         {
            if (G_ReloadLevel())
                return true;
```

**Comment by JNechaevsky** [comment]
Nope! Now it's just warps to the next level. I've also checked it in another PC - same here, it's does not seems to be a keyboard bug.

**Comment by fabiangreffrath** [comment]
> Now it's just warps to the next level

That's it! For keys without a default binding I have to add another check:

```
--- a/src/doom/m_menu.c
+++ b/src/doom/m_menu.c
@@ -2918,12 +2918,12 @@ boolean M_Responder (event_t* ev)
        }
         // [crispy] those two can be considered as shortcuts for the IDCLEV cheat
         // and should be treated as such, i.e. add "if (!netgame)"
-        else if (!netgame && key == key_menu_reloadlevel)
+        else if (!netgame && key != 0 && key == key_menu_reloadlevel)
         {
            if (G_ReloadLevel())
                return true;
         }
-        else if (!netgame && key == key_menu_nextlevel)
+        else if (!netgame && key != 0 && key == key_menu_nextlevel)
         {
            if (G_GotoNextLevel())
                return true;
```

**Comment by JNechaevsky** [comment]
Okay, now I have no level jumps, and more of that, I've tracked down what was happening. 

So the problem is - if I have only binded "clean screenshot" (i.e. not common "screenshot) key, this will break things. After applying your last patch there are no more level jumping, but if common "screenshot" is not binded, nothing is happening after pressing binded "clean screenshot", only menu breaking.

In other words: your patch is working, but no need to apply it, the problem with another thing!

Could you please try by your side: 
- Remove bindings for both screenshot types
- Bind only "clean screenshot" for some kind of key
- Start Crispy and press that binded key
- Press `ESC` and see what will happen
